Fix TreeView missing children issue

Felix-Dev opened the pull request:
## Description
With data bound TreeView items some recycled item containers (TreeViewItem) are losing their bound children collection. Futher debugging the code shows that the TreeView nodes of such TreeViewItems also have lost their children (TreeViewNode.m_itemsDataSource is null). As such, calling TreeViewNode.SyncChildrenNodesWithItemsSource() is not fixing this issue.

The current fix is a first attempt at fixing the issue. It forces a binding re-evaluation of the TreeViewItem.ItemsSource dependency property. As such, TreeViewItem.OnPropertyChanged is called which in return sets the items source of the corresponding TreeViewNode (TreeViewNode.ItemsSource(itemsSource)).

As we are basically doing the steps after TreeView creation again, there might be a better way to fix the issue. I would like to ask a team help aid me in search for a better fix in case there is one 🙂

## Motivation and Context
Fixes https://github.com/microsoft/microsoft-ui-xaml/issues/1701 and Fixes https://github.com/microsoft/microsoft-ui-xaml/issues/523.

## How Has This Been Tested?
Added test to the MUXControlsTestApp TreeView test page. Test can be modified if requested.

japf commented on dev/TreeView/TestUI/TreeViewPage.xaml.cs:
```diff
@@ -998,5 +998,20 @@ private void ExpandRootNode_Click(object sender, RoutedEventArgs e)
+        private bool isTreeView1SourceInUse = true;
+        private void SwapItemsSource_Click(object sender, RoutedEventArgs e)
+        {
+            if (!isTreeView1SourceInUse)
```
You can maybe drop the boolean field and use `if (this.ContentModeTestTreeView.ItemsSource == TestTreeView2ItemsSource)` instead?

Felix-Dev replied:
Updated :)

japf commented:
I was looking into this exact issue yesterday, fun timing :)

ranjeshj commented on dev/TreeView/TestUI/TreeViewPage.xaml:
```diff
@@ -74,6 +74,7 @@
                     <Button x:Name="ResetItemsSource" AutomationProperties.Name="ResetItemsSource" Content="ResetItemsSource" HorizontalAlignment="Stretch" Margin="1" Click="ResetItemsSource_Click"/>
                     <Button x:Name="ResetItemsSourceAsync" AutomationProperties.Name="ResetItemsSourceAsync" Content="ResetItemsSourceAsync" HorizontalAlignment="Stretch" Margin="1" Click="ResetItemsSourceAsync_Click"/>
                     <Button x:Name="ExpandRootNode" AutomationProperties.Name="ExpandRootNode" Content="ExpandRootNode" HorizontalAlignment="Stretch" Margin="1" Click="ExpandRootNode_Click"/>
+                    <Button x:Name="SwapItemsSource" AutomationProperties.Name="SwapItemsSource" Content="SwapItemsSource" HorizontalAlignment="Stretch" Margin="1" Click="SwapItemsSource_Click"/>
```
Can you please add an interaction test exercising this scenario ?

kaiguo replied:
It's a flag to balancing the workload in test pipeline, TestSuiteB should be fine.

ranjeshj commented:
/azp run

kaiguo commented on dev/TreeView/TreeViewList.cpp:
```diff
@@ -99,6 +99,14 @@ void TreeViewList::OnContainerContentChanging(const winrt::IInspectable& /*sende
+        // "Re-evaluate" the item source binding to trigger a TreeViewItem.OnPropertyChanged(ItemsSourceProperty) call.
+        if (auto itemsSource = targetItem.ItemsSource())
+        {
+            targetItem.ItemsSource(nullptr);
```
What happens if you don't set it to nullptr?

Felix-Dev replied:
The current fix is re-doing the steps after a TreeView is created . I'm setting the source to null so that setting the source again will trigger a TreeViewItem::OnPropertyChanged call. This call is then setting up the child nodes (as TreeViewNode.m_itemsDataSource also seems to be lost upon recycling and needs to be set again). If I don't set it to nullptr here, I would have to set it to something else to trigger the OnPropertyChanged call.

There might be a better way to fix this issue (and I'm also seeing some tests failing locally for me). For now I only found this attempt though.

kaiguo replied:
I see, you can probably wrap the [code](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/TreeView/TreeViewItem.cpp#L327-L334) in OnPropertyChanged into a separate function and call that function instead, that way you don't have to set ItemsSource to null first just to trigger OnPropertyChanged.

Felix-Dev replied:
Done.

japf commented on dev/TreeView/TreeViewList.cpp:
```diff
@@ -99,6 +99,14 @@ void TreeViewList::OnContainerContentChanging(const winrt::IInspectable& /*sende
         auto targetItem = args.ItemContainer().as<winrt::TreeViewItem>();
+
+        // "Re-evaluate" the item source binding to trigger a TreeViewItem.OnPropertChanged(ItemsSourceProperty) changed call.
+        if (auto itemsSource = targetItem.ItemsSource())
```
Maybe we could rename this new method to `UpdateItemsSource` to be consistent with the `UpdateIndentation` and `UpdatesSelectionVisual` that already exist.

Felix-Dev replied:
Already done in commit https://github.com/microsoft/microsoft-ui-xaml/pull/2058/commits/d4a1f079982f129f679be35043d902ae0dd914a8 🙂 New function is named "SetItemsSource" for now but I was also thinking about "UpdateItemsSource" for the name. TreeViewItem also has functions to get/set the ItemsSource (TreeViewItem.ItemSource()) so introducing a new function with "Set" in the name might be confusing (there also is no "GetItemsSource"). Will update name to "UpdateItemsSource".

chingucoding replied:
I agree with @japf here, I think it may be better to make the function public and call it, instead of relying on the property changed event to call the function.

Felix-Dev replied:
Already done in commit https://github.com/microsoft/microsoft-ui-xaml/commit/d4a1f079982f129f679be35043d902ae0dd914a8 🙂

japf replied:
I'm wondering if this API (`SetItemsSource`) needs to be public. From the "outside", when using the tree view control you should not need to deal with the underlying `TreeViewList` control. That looks like an implementation details to me.

Felix-Dev replied:
TreeViewItem needs to expose this function so that it can be called by TreeViewList. You don't have access to this function outside of WinUI (i.e. a project consuming WinUI) as it is not part of the IDL file. Can you clarify what you mean here?

japf replied:
You're right, if that's not part of the IDL that's fine :-)
Thanks for clarifying.

kaiguo commented:
/azp run

StephenLPeters commented:
@Felix-Dev the test failure looks like its a real issue to me, are you able to look into it?

Felix-Dev commented:
@StephenLPeters I'm seeing the test failures locally too. I will look at them this weekend and see if there is something I can do.

StephenLPeters commented:
@kaiguo thoughts on @Felix-Dev s comment?

kaiguo commented:
@Felix-Dev I think It's because you are trying to update ListView ItemsSource during layout pass. The node expansion operation is basically adding all the children into Listview ItemsSource which triggers a layout pass and calls ContainerContantChanging, and inside ContainerContantChanging ItemsSource is being set again, it'll trigger another layout pass. @ranjeshj are there other places we can safely put the ItemsSource updating code?

ranjeshj commented:
> @Felix-Dev I think It's because you are trying to update ListView ItemsSource during layout pass. The node expansion operation is basically adding all the children into Listview ItemsSource which triggers a layout pass and calls ContainerContantChanging, and inside ContainerContantChanging ItemsSource is being set again, it'll trigger another layout pass. @ranjeshj are there other places we can safely put the ItemsSource updating code?

Please avoid updating ItemsSource (or changing the collection in any way) within layout. Doing so will likely end up causing issues down the line. You can set it through a dispatcher call though which will run after layout completes.

Felix-Dev commented:
@ranjeshj @kaiguo So I added a dispatcher call (looked at TreeViewList::PrepareContainerForItemOverride) and it added a bunch of new problems:
```cs
if (auto itemsSource = targetItem.ItemsSource())
{
    const DispatcherHelper dispatcher{ *this };
        dispatcher.RunAsync(
            [treeViewItem, targetNode, itemsSource]()
            {
                treeViewItem->SetItemsSource(targetNode, itemsSource);
            });
}   
```
See the following image for the failing test case for example (collapse Root with both Root.0 and Root.1 collapsed, then expand):
![image](https://user-images.githubusercontent.com/1398851/76995234-d5e35600-694f-11ea-97f8-ab0f8adac9b6.png)
with expected look:
![image](https://user-images.githubusercontent.com/1398851/76995293-f27f8e00-694f-11ea-9c8c-05a69e186fce.png)

If I now expand Root.2 then collapse Root and expand Root again, the app crashes as in the previous cases (only Root0 substituted with Root2).

Ignore the dispatcher code for now, I added another check to the code which seems to fix the crash and failing tests just fine:
```cs
if (auto itemsSource = targetItem.ItemsSource())
{
    if (treeViewNode->ItemsSource() == nullptr)
    {
        treeViewItem->SetItemsSource(targetNode, itemsSource);
    }
}  
```
This is based on the observation that the treeview issues this PR attempts to fix really only happene dbecause TreeViewNode::m_itemDataSource was null for the failing treeview elements. Now this check makes sure we only re-set the item source where there's a mismatch betwee the container and the underlying tree node (mismatch: container has an item source whereas treenode does not).

Since this check fixes the issues I then wrapped it inside a dispatcher call. That, however, caused three new tests to fail: ItemsSourceResycTest, TreeViewRootNodeBindingTest and TreeViewWithMultiLevelChildrenExpandCollapseTest.

I decided to push the new check for now as that makes all my local tests run fine. If the tests run fine here too, we can see how to put that in a dispatcher call if still required so that no new test failures will occur.

ranjeshj commented:
/azp run

Felix-Dev commented:
@kaiguo @ranjeshj Now that all tests have succeeded and the call is not yet wrapped inside a dispatcher call (which would cause new tets failures based on the attempt above), how do we want to continue here?

kaiguo commented:
> @kaiguo Kai Guo FTE @ranjeshj Ranjesh Jaganathan FTE Now that all tests have succeeded and the call is not yet wrapped inside a dispatcher call (which would cause new test failures based on the attempt above), how do we want to continue here?

If null check fixes the problem I think we should go with the null check approach since dispatcher call makes the code hard to debug and understand, but I'll defer to @ranjeshj to make the call : )

ranjeshj commented:
> > @kaiguo Kai Guo FTE @ranjeshj Ranjesh Jaganathan FTE Now that all tests have succeeded and the call is not yet wrapped inside a dispatcher call (which would cause new test failures based on the attempt above), how do we want to continue here?
> 
> If null check fixes the problem I think we should go with the null check approach since dispatcher call makes the code hard to debug and understand, but I'll defer to @ranjeshj to make the call : )

Ok. The null check sounds reasonable and we can revisit if this becomes an issue.
